Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/utilities/RegExpUtility.java

```diff
@@ -177,14 +174,8 @@ public static Pattern getSafeLookbehindRegExp(String source) {
     }
 
     public static Pattern getSafeLookbehindRegExp(String source, int flags) {
-
         String result = source;
-
-        // Java pre 1.9 doesn't support unbounded lookbehind lengths
-        if (unboundedLookBehindNotSupported) {

```

Comment:
Great to see that calling this fallback method from those two regexes passes all tests in Java 8.
But we should not use the fallback in versions where the declared regex works as is. Using this fallback may lead to perf issues and to side effects in the future.
Can you re-enable the version check that you folks had fixed before? This should properly fix the issue.
Also, why didn't this PR enable the tests Lionbridge disabled in that other PR (mentioned on this issue/PR discussion)? Whatever fix comes from this, should apply to those test cases as well.

Reply:
PR #2383. Spec cases marked as: "Comment": "Java does not correctly handle lookbehinds."
should be enabled and the comment removed.

Reply:
@tellarin, after re-enabling the version check, the unit tests are failing using Java 11. That's why we unified the fallback method for all the JDKs.

We will be working on adding re-enabled specs which were disabled in the PR #2383 in order to add them in this PR.

_5 specs are failing using Java 11_
![image](https://user-images.githubusercontent.com/11904023/106601538-86dbea80-653a-11eb-9ef1-3cdca167b5cc.png)

Reply:
@tellarin, we confirmed that re-enabling the mentioned specs of #2383 and incorporating the JDK's validation, the specs are failing no matter which version of Java is used (e.g. Java 8, Java 11).

Also, we found the same problem in JS rollbacking the changes of PR #1069. In that PR, NodeJS version 10.3.0 starts supporting Lookbehind Assertions so all the machinery introduced in RegExpUtility to deal with them was unnecesary.

We tried to apply the same changes, however, a lot of specs were failing. This means that Java doesn't support those spec cases.

Finally, we think that it's necessary to unify the fallback method and the specs are correctly disabled currently. Are you okay with this?

Reply:
This is interesting. The current build passes, so the unbouded look-behind is supported in Java (I guess the build is Java 15). We should check in which Java version this unbounded issue was resolved and that version check to trigger the fallback should reflect that.
Are the other cases from #2383 failing due to the same issue? Or a different look-behind issue?

Reply:
Hi @tellarin, the current build passed as it's using the fallback method by default for all the JDKs in this PR.

We **summarized** all the scenarios in the following list:
1. Using `master` branch
    * ❌Java 8: 62 tests are broken (all for look-behind's issue)
    * ✔Java 11: all tests passed
    * ✔Java 15: all tests passed
2. Using `master` branch re-enabling the disabled tests of PR #2383
    * ❌Java 8: 66 tests are broken (62 for look-behind's issue, 4 assertionerror)
    * ❌Java 11: 4 tests are broken (all for assertionerror), the four are the re-enabled
    * ❌Java 15: 4 tests are broken (all for assertionerror), the four are the re-enabled
3. Branch of PR #2439 with JDKs validation
    * ❌Java 8: 62 tests are broken (all for look-behind's issue)
    * ❌Java 11: 5 tests are broken (all for assertionerror)
    * ❌Java 15: 5 tests are broken (all for assertionerror)
4. Branch of PR #2439 with JDKs validation and re-enabling the disabled tests of PR #2383
    * ❌Java 8: 66 tests are broken (62 for look-behind's issue, 4 assertionerror)
    * ❌Java 11: 9 tests are broken (all for assertionerror), including the re-enabled
    * ❌Java 15: 9 tests are broken (all for assertionerror), including the re-enabled
5. Branch of PR #2439 without JDKs validation (unifying the fallback method)
    * ✔Java 8: all tests passed
    * ✔Java 11: all tests passed
    * ✔Java 15: all tests passed
5. Branch of PR #2439 without JDKs validation (unifying the fallback method) and re-enabling the disabled tests of PR #2383
    * ❌Java 8: 4 tests are broken (all for assertionerror), the four are the re-enabled
    * ❌Java 11: 4 tests are broken (all for assertionerror), the four are the re-enabled
    * ❌Java 15: 4 tests are broken (all for assertionerror), the four are the re-enabled

We came up with 2 **conclusions**:
1. As mentioned above, we think that Java doesn't support lookbehind fully and it's necessary to use the fallback method no matter which version of Java is used 
2. The disabled specs are correct to be disabled

Let us know what do you think about this 😊

Reply:
So it seems the disabled cases in #2383 fail anyways. Any clues on what's the root cause of those 4 failures? The main goal of improving this fallback behaviour is to have a solution that also works for them.
If this is a complex item it could come in its own PR. 
Also, according to your description above, if we have a separate PR for that, the right solution for this one (fixing the current failing tests and the java version check, wihtout re-enabling these four spec cases) if to include your fix to the JDK version check so the fallback only applies to Java 8 and call the method for those two NumberRange regexes. 
Your list item 1) shows that's the only situation where things were failing.

Reply:
@tellarin, thanks for your comment. Your understanding is **right**, we think that we can **merge** this PR, then do the correct JDK validation and fix the broken specs, including the disabled ones of PR #2383.

> I see. So Java 11 and 15 were only working because the fallback was wrongly triggered, right? I.e. the "unbounded lookbehind" issue afects all versions?

Exactly, Java 11 and Java 15 were only working because the fallback method was wrongly triggered. With the correct validation, it will fail some specs (5 for assertionerror)

> And also those 4 cases from PR #2383 fail due to other lookbehind issue Java?

The problem is not a lookbehind issue but an assertion error. Currently, we don't know the root cause of the problem and we think that it would be too complex to address it right now.